File: go/test/moon_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/voxgig-solardemo-sdk/go"
	"github.com/voxgig-sdk/voxgig-solardemo-sdk/go/core"

	vs "github.com/voxgig-sdk/voxgig-solardemo-sdk/go/utility/struct"
)

func TestMoonEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Moon(nil)
		if ent == nil {
			t.Fatal("expected non-nil MoonEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"moon": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Moon(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.MakeConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Moon(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := moonBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "moon." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set VOXGIGSOLARDEMO_TEST_MOON_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		moonRef01Ent := client.Moon(nil)
		moonRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "moon"}, setup.data), "moon_ref01"))
		moonRef01Data["planet_id"] = setup.idmap["planet01"]

		moonRef01DataResult, err := moonRef01Ent.Create(moonRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		moonRef01Data = core.ToMapAny(moonRef01DataResult)
		if moonRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if moonRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		moonRef01Match := map[string]any{
			"planet_id": setup.idmap["planet01"],
		}

		moonRef01ListResult, err := moonRef01Ent.List(moonRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		moonRef01List, moonRef01ListOk := moonRef01ListResult.([]any)
		if !moonRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", moonRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(moonRef01List), map[string]any{"id": moonRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		moonRef01DataUp0Up := map[string]any{
			"id": moonRef01Data["id"],
			"planet_id": setup.idmap["planet_id"],
		}

		moonRef01MarkdefUp0Name := "kind"
		moonRef01MarkdefUp0Value := fmt.Sprintf("Mark01-moon_ref01_%d", setup.now)
		moonRef01DataUp0Up[moonRef01MarkdefUp0Name] = moonRef01MarkdefUp0Value

		moonRef01ResdataUp0Result, err := moonRef01Ent.Update(moonRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		moonRef01ResdataUp0 := core.ToMapAny(moonRef01ResdataUp0Result)
		if moonRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if moonRef01ResdataUp0["id"] != moonRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if moonRef01ResdataUp0[moonRef01MarkdefUp0Name] != moonRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", moonRef01MarkdefUp0Name, moonRef01ResdataUp0[moonRef01MarkdefUp0Name])
		}

		// LOAD
		moonRef01MatchDt0 := map[string]any{
			"id": moonRef01Data["id"],
		}
		moonRef01DataDt0Loaded, err := moonRef01Ent.Load(moonRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		moonRef01DataDt0LoadResult := core.ToMapAny(moonRef01DataDt0Loaded)
		if moonRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if moonRef01DataDt0LoadResult["id"] != moonRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		moonRef01MatchRm0 := map[string]any{
			"id": moonRef01Data["id"],
		}
		_, err = moonRef01Ent.Remove(moonRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		moonRef01MatchRt0 := map[string]any{
			"planet_id": setup.idmap["planet01"],
		}

		moonRef01ListRt0Result, err := moonRef01Ent.List(moonRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		moonRef01ListRt0, moonRef01ListRt0Ok := moonRef01ListRt0Result.([]any)
		if !moonRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", moonRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(moonRef01ListRt0), map[string]any{"id": moonRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func moonBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "moon", "MoonTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read moon test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse moon test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"moon01", "moon02", "moon03", "planet01", "planet02", "planet03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("VOXGIGSOLARDEMO_TEST_MOON_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"VOXGIGSOLARDEMO_TEST_MOON_ENTID": idmap,
		"VOXGIGSOLARDEMO_TEST_LIVE":      "FALSE",
		"VOXGIGSOLARDEMO_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["VOXGIGSOLARDEMO_TEST_MOON_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add planet_id alias for update test.
	if idmapResolved["planet_id"] == nil {
		idmapResolved["planet_id"] = idmapResolved["planet01"]
	}

	if env["VOXGIGSOLARDEMO_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
			},
			extra,
		})
		client = sdk.NewVoxgigSolardemoSDK(core.ToMapAny(mergedOpts))
	}

	live := env["VOXGIGSOLARDEMO_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["VOXGIGSOLARDEMO_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: cpp/test/moon_entity_test.cpp
// Generated basic-flow test for the moon entity (model-driven,
// unit mode; mirrors the rust/go TestEntity generator).

#include "runner_support.hpp"

using namespace sdk;
using namespace sdk::rs;

struct MoonSetup {
  std::shared_ptr<SolardemoSDK> client;
  Value data;
  Value idmap;
  Value env;
  bool live = false;
  bool synthetic_only = false;
  long long now = 0;
};

static MoonSetup moon_basic_setup(const Value& extra) {
  load_env_local();

  std::string entity_data_file = "../.sdk/test/entity/moon/MoonTestData.json";
  Value entity_data = vs::parse_json(read_file(entity_data_file));

  Value options = vmap({{"entity", getp(entity_data, "existing")}});
  auto client = SolardemoSDK::testSDK(options, extra);

  // idmap via transform (upper-cased id name synthetics), matching the donors.
  Value idmap = Struct::transform(
      vlist({Value("moon01"), Value("moon02"), Value("moon03"), Value("planet01"), Value("planet02"), Value("planet03")}),
      vmap({{"`$PACK`", vlist({
        Value(""),
        vmap({
          {"`$KEY`", Value("`$COPY`")},
          {"`$VAL`", vlist({Value("`$FORMAT`"), Value("upper"), Value("`$COPY`")})}
        })
      })}}));
  if (!idmap.is_map()) idmap = vmap();

  Value env = env_override(vmap({
    {"SOLARDEMO_TEST_MOON_ENTID", idmap},
    {"SOLARDEMO_TEST_LIVE", Value("FALSE")},
    {"SOLARDEMO_TEST_EXPLAIN", Value("FALSE")}
  }));

  Value idmap_resolved = Helpers::toMapAny(getp(env, "SOLARDEMO_TEST_MOON_ENTID"));
  if (!idmap_resolved.is_map()) idmap_resolved = idmap;

  if (getp(idmap_resolved, "planet_id").is_undef()) {
    setp(idmap_resolved, "planet_id", getp(idmap_resolved, "planet01"));
  }

  bool live = getp(env, "SOLARDEMO_TEST_LIVE") == Value("TRUE");

  MoonSetup s;
  s.client = client;
  s.data = entity_data;
  s.idmap = idmap_resolved;
  s.env = env;
  s.live = live;
  s.synthetic_only = false;
  s.now = now_ms();
  return s;
}

static void moon_entity_instance() {
  auto testsdk = SolardemoSDK::testSDK();
  auto ent = testsdk->moon();
  ASSERT_EQ(ent->getName(), std::string("moon"), "entity name");
}


static void moon_entity_stream() {
  // stream() runs the list op through the full pipeline and returns the
  // result items. Seed two entities via test mode; with the streaming feature
  // active it yields the feature's incremental items, else it falls back to
  // the materialised items — either way every item is yielded.
  Value seed = vmap({{"entity", vmap({{"moon", vmap({
      {"strm01", vmap({{"id", Value("strm01")}})},
      {"strm02", vmap({{"id", Value("strm02")}})}})}})}});
  Value sdkopts = vmap({{"feature",
      vmap({{"streaming", vmap({{"active", Value(true)}})}})}});

  auto strsdk = SolardemoSDK::testSDK(seed, sdkopts);
  auto se = strsdk->moon();
  std::vector<Value> items = se->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)items.size(), 2, "stream yields both seeded items");

  auto plainsdk = SolardemoSDK::testSDK(seed, Value::undef());
  auto pe = plainsdk->moon();
  std::vector<Value> pitems = pe->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)pitems.size(), 2, "fallback stream yields both items");
}

static void moon_entity_basic() {
  auto setup = moon_basic_setup(Value::undef());
  std::string mode = setup.live ? "live" : "unit";
  for (const std::string& op : std::vector<std::string>{"create", "list", "update", "load", "remove"}) {
    auto sk = is_control_skipped("entityOp", std::string("moon.") + op, mode);
    if (sk.first) { std::cerr << "skip: " << (sk.second.empty()? "sdk-test-control.json" : sk.second) << "\n"; return; }
  }
  auto client = setup.client;
  // CREATE
  auto moon_ref01_ent = client->moon();
  Value moon_ref01_data = Helpers::toMapAny(getp(Struct::getpath(setup.data, {"new", "moon"}), "moon_ref01"));
  if (!moon_ref01_data.is_map()) moon_ref01_data = vmap();
  setp(moon_ref01_data, "planet_id", getp(setup.idmap, "planet01"));
  {
    Value moon_ref01_data_result = moon_ref01_ent->create(Struct::clone(moon_ref01_data), Value::undef())->data();
    moon_ref01_data = Helpers::toMapAny(moon_ref01_data_result);
    if (!moon_ref01_data.is_map()) moon_ref01_data = vmap();
    ASSERT_TRUE(moon_ref01_data.is_map(), "expected create result to be a map");
    ASSERT_TRUE(!getp(moon_ref01_data, "id").is_undef(), "expected created entity to have an id");
  }

  // LIST
  Value moon_ref01_match = vmap();
  setp(moon_ref01_match, "planet_id", getp(setup.idmap, "planet01"));
  auto moon_ref01_list_ents = moon_ref01_ent->list(Struct::clone(moon_ref01_match), Value::undef());
  // list resolves to one ENTITY per record; the flow asserts on the records.
  Value moon_ref01_list = vlist();
  for (const auto& e : moon_ref01_list_ents) { moon_ref01_list.as_list()->push_back(e->data()); }
  ASSERT_TRUE(moon_ref01_list.is_list(), "expected list result to be an array");
  {
    std::vector<Value> found = Struct::select(entity_list_to_data(moon_ref01_list), vmap({{"id", getp(moon_ref01_data, "id")}}));
    ASSERT_TRUE(!found.empty(), "expected to find created entity in list");
  }

  // UPDATE
  Value moon_ref01_data_up0_up = vmap();
  setp(moon_ref01_data_up0_up, "id", getp(moon_ref01_data, "id"));
  setp(moon_ref01_data_up0_up, "planet_id", getp(setup.idmap, "planet_id"));
  std::string moon_ref01_data_up0_markval = std::string("Mark01-moon_ref01_") + std::to_string(setup.now);
  setp(moon_ref01_data_up0_up, "kind", Value(moon_ref01_data_up0_markval));
  Value moon_ref01_resdata_up0_result = moon_ref01_ent->update(Struct::clone(moon_ref01_data_up0_up), Value::undef())->data();
  Value moon_ref01_resdata_up0 = Helpers::toMapAny(moon_ref01_resdata_up0_result);
  if (!moon_ref01_resdata_up0.is_map()) moon_ref01_resdata_up0 = vmap();
  ASSERT_TRUE(moon_ref01_resdata_up0.is_map(), "expected update result to be a map");
  ASSERT_EQ_VAL(getp(moon_ref01_resdata_up0, "id"), getp(moon_ref01_data_up0_up, "id"), "expected update result id to match");
  ASSERT_EQ_VAL(getp(moon_ref01_resdata_up0, "kind"), Value(moon_ref01_data_up0_markval), "expected kind to be updated");

  // LOAD
  Value moon_ref01_match_dt0 = vmap({{"id", getp(moon_ref01_data, "id")}});
  Value moon_ref01_data_dt0_loaded = moon_ref01_ent->load(Struct::clone(moon_ref01_match_dt0), Value::undef())->data();
  Value moon_ref01_data_dt0_load_result = Helpers::toMapAny(moon_ref01_data_dt0_loaded);
  ASSERT_TRUE(moon_ref01_data_dt0_load_result.is_map(), "expected load result to be a map");
  ASSERT_EQ_VAL(getp(moon_ref01_data_dt0_load_result, "id"), getp(moon_ref01_data, "id"), "expected load result id to match");

  // REMOVE
  {
    Value moon_ref01_match_rm0 = vmap({{"id", getp(moon_ref01_data, "id")}});
    moon_ref01_ent->remove(Struct::clone(moon_ref01_match_rm0), Value::undef());
  }

  // LIST
  Value moon_ref01_match_rt0 = vmap();
  setp(moon_ref01_match_rt0, "planet_id", getp(setup.idmap, "planet01"));
  auto moon_ref01_list_rt0_ents = moon_ref01_ent->list(Struct::clone(moon_ref01_match_rt0), Value::undef());
  // list resolves to one ENTITY per record; the flow asserts on the records.
  Value moon_ref01_list_rt0 = vlist();
  for (const auto& e : moon_ref01_list_rt0_ents) { moon_ref01_list_rt0.as_list()->push_back(e->data()); }
  ASSERT_TRUE(moon_ref01_list_rt0.is_list(), "expected list result to be an array");
  {
    std::vector<Value> found = Struct::select(entity_list_to_data(moon_ref01_list_rt0), vmap({{"id", getp(moon_ref01_data, "id")}}));
    ASSERT_TRUE(found.empty(), "expected removed entity to not be in list");
  }

}

int main() {
  T_RUN(moon_entity_instance);
  T_RUN(moon_entity_stream);
  T_RUN(moon_entity_basic);
  return sdktest::summary("moon_entity_test");
}
